Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,354 @@
+
+
+## High-level proposal
+In .NET Core 3.0 the hosting layer (see [here](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md)) ships with several hosts. These are binaries which act as the entry points to the .NET Core hosting/runtime:
+* The "muxer" (`dotnet.exe`)
+* The `apphost` (`.exe` which is part of the app)
+* The `comhost` (`.dll` which is part of the app and acts as COM server)
+* The `ijwhost` (`.dll` consumed via `.lib` used by IJW assemblies)
+
+Every one of these hosts serve different scenario and expose different APIs. The one thing they have in common is that their main purpose is to find the right `hostfxr`, load it and call into it to execute the desired scenario. For the most part all these hosts are basically just wrappers around functionality provided by `hostfxr`.
+
+The proposal is to add a new host library `nethost` which can be used by native host to easily host managed components and to easily locate `hostfxr` for more advanced scenarios.
+
+At the same time add the ability to pass additional runtime properties when starting the runtime through the hosting entry points (starting app, loading component). This can be used by the native host to:
+* Register startup hook without modifying environment variables (which are inherited by child processes)
+* Introduce new runtime knobs which are only available for native hosts without the need to update the hosting APIs every time.
+
+
+*Technical note: All strings in the proposed APIs are using the `char_t` in this document for simplicity. In real implementation they are of the type `pal::char_t`. In particular:*
+* *On Windows - they are `WCHAR *` using `UTF16` encoding*
+* *On Linux/macOS - they are `char *` using `UTF8` encoding*
+
+
+## New host binary for finding `hostfxr`
+Add new library `nethost` which will provide a way to locate the right `hostfxr`.
+The library would be a dynamically loaded library (`.dll`, `.so`, `.dylib`). For ease of use there would be a header file for C/C++ apps as well as `.lib`/`.a` for easy linking.
+Native host would ship this library as part of the app. Unlike the `apphost`, `comhost` and `ijwhost`, the `nethost` will not be directly supported by the .NET Core SDK since it's target usage is not from .NET Core apps.
+
+The exact delivery mechanism is TBD (pending investigation), but it probably should include NuGet (for C++ projects) and plain `.zip` (for any consumer). The binary itself should be signed by Microsoft as there will be no support for modifying the binary as part of custom application build (unlike `apphost`).
+
+
+### Locate `hostfxr`
+``` C++
+int nethost_get_hostfxr_path(
```

Comment:
Will 'nethost_' be part of the name?  Or only there to help differentiate which binary it is an export of?

Reply:
This was more for consistency - all the hosting libraries prefix the exports with the library name (`hostfxr_..`, `corehost_..`, …).
I guess since this is the "easy to use" library, we may omit the prefix for simplicity.